### PR TITLE
fix(cluster-autoscaler): prevent panic in SimulateNodeRemoval by handling missing node info

### PR DESCRIPTION
**Pull request by kincoy**
#### What type of PR is this?

/kind bug

---

#### What this PR does / why we need it:

This PR fixes a potential `nil` dereference issue in `SimulateNodeRemoval` when a node is missing from the `clusterSnapshot`.

Previously, if `clusterSnapshot.GetNodeInfo` failed, the function would continue and potentially panic when accessing `nodeInfo.Pods()`.

This fix introduces:
- A new `UnremovableReason`: `NoNodeInfo`, used to mark such nodes as unremovable.
- An early return from `SimulateNodeRemoval` when the node is missing.
- A defensive placeholder node (with only `.Name` set) to maintain observability and event compatibility.
- A dedicated unit test case to verify this scenario is correctly handled.

---

#### Which issue(s) this PR fixes:

N/A

---

#### Special notes for your reviewer:

This PR fixes a potential nil dereference in `SimulateNodeRemoval` when a node is missing from the cluster snapshot.  
It adds a new `UnremovableReason` (`NoNodeInfo`) to capture this edge case, and add the test coverage.

---

#### Does this PR introduce a user-facing change?
```release-note
fix: handle missing node info in SimulateNodeRemoval
```

### CODE REVIEW

**Review by elmiko — COMMENTED** [review]
thank you for the PR and adding the test too.

the changes here look generally good to me, but i am not overly familiar with the simulator code. in specific, the returns for the failure condition.

would it be possible to add a unit test for the other return value as well? (i see the test for the unremovable node with the `NoNodeInfo`, but can we also test for `UnexpectedError`?)

**Comment by kincoy** [comment]
> thank you for the PR and adding the test too.
> 
> the changes here look generally good to me, but i am not overly familiar with the simulator code. in specific, the returns for the failure condition.
> 
> would it be possible to add a unit test for the other return value as well? (i see the test for the unremovable node with the `NoNodeInfo`, but can we also test for `UnexpectedError`?)

Thanks! I looked into `GetNodeInfo` — aside from `ErrNodeNotFound`, other errors only happen when `draEnabled` is true and `WrapSchedulerNodeInfo` fails, which is rare and hard to simulate without artificial mocks.

If you have a clean way to test this case, I’m happy to give it a try!

**Comment by elmiko** [comment]
> Thanks! I looked into GetNodeInfo — aside from ErrNodeNotFound, other errors only happen when draEnabled is true and WrapSchedulerNodeInfo fails, which is rare and hard to simulate without artificial mocks.

thanks for investigating, i was worried it might take a complicated mock to make it work. i don't think it's worth the effort to create a test with a mock just for the error condition.

/lgtm

would be good to get another review from a maintainer

File: cluster-autoscaler/simulator/cluster.go [hunk 1]
```diff
@@ -89,6 +91,8 @@ const (
 	NoPlaceToMovePods
 	// BlockedByPod - node can't be removed because a pod running on it can't be moved. The reason why should be in BlockingPod.
 	BlockedByPod
+	// NoNodeInfo - node can't be removed because it doesn't have any node info in the cluster snapshot.
+	NoNodeInfo
```
Comment (jackfrancis) [inline cluster-autoscaler/simulator/cluster.go]: We need to move this new const iota member to the end of the list, otherwise `UnexpectedError` will be changed from `16` to `17`. (there should be a lint rule for this...)

File: cluster-autoscaler/simulator/cluster.go [hunk 2]
```diff
@@ -151,6 +155,11 @@ func (r *RemovalSimulator) SimulateNodeRemoval(
 	nodeInfo, err := r.clusterSnapshot.GetNodeInfo(nodeName)
 	if err != nil {
 		klog.Errorf("Can't retrieve node %s from snapshot, err: %v", nodeName, err)
+		ghostNode := &apiv1.Node{ObjectMeta: metav1.ObjectMeta{Name: nodeName}}
```
Comment (jackfrancis) [inline cluster-autoscaler/simulator/cluster.go]: I have a slight preference for doing something like this

```go
		unremovableReason := UnexpectedError
		if errors.Is(err, clustersnapshot.ErrNodeNotFound) {
			unremovableReason = NoNodeInfo
		}
		unremovableNode := &UnremovableNode{Node: &apiv1.Node{ObjectMeta: metav1.ObjectMeta{Name: nodeName}}, Reason: unremovableReason}
		return nil, unremovableNode
```

1) to avoid the multiple return statements
2) to actuate what the the if condition is really trying to determine
3) to avoid wrapping the node object in an "opinonated" variable name like "ghostNode", which may indicate something more interesting going on than actually is

File: cluster-autoscaler/simulator/cluster.go (hunk 1 above)
Comment (kincoy) [reply cluster-autoscaler/simulator/cluster.go]: Done — I’ve moved `NoNodeInfo` to the end of the list to preserve existing iota values.

File: cluster-autoscaler/simulator/cluster.go (hunk 2 above)
Comment (kincoy) [reply cluster-autoscaler/simulator/cluster.go]: Makes sense — updated as you suggested. Also applied consistent naming in both implementation and tests.Thanks for the helpful review!

**Comment by kincoy** [comment]
Friendly ping @jackfrancis  — this PR has been idle for a while. Would appreciate a review when convenient 🙏

**Comment by jackfrancis** [comment]
@kincoy sorry for the delay, this lgtm, and I'll give it the official stamp after a rebase

thanks!

**Comment by kincoy** [comment]
@jackfrancis Thanks for your review. I've resolved the merge conflict in the PR. Could you please re-apply the LGTM label?
Also, I noticed that the conflict was caused by the deletion of the cluster_test.go file. Would you mind sharing the reason for removing this file? Thanks!

**Comment by jackfrancis** [comment]
/release-note-edit
```release-note
fix: handle missing node info in SimulateNodeRemoval
```

**Comment by jackfrancis** [comment]
/lgtm
/approve
/hold

@towca @BigDarkClown @x13n @elmiko for another review

**Comment by x13n** [comment]
Hm... so https://github.com/kubernetes/autoscaler/pull/8473 removed unit tests along with the tested function, but the underlying `SimulateNodeRemoval` was only indirectly tested. In retrospect, the tests should've been probably rewritten instead of removed, so that this PR could extend them. @jackfrancis do you think you could bring the tests back, updating them to test `SimulateNodeRemoval` instead of `FindNodesToRemove` that got deleted? Alternatively, @kincoy would you be able to do that as a part of this PR?

**Comment by kincoy** [comment]
> @kincoy would you be able to do that as a part of this PR?

Yeah, I’ll add a new test file in this PR to cover `SimulateNodeRemoval`.

**Comment by kincoy** [comment]
@x13n hi, I’ve updated the PR and added the test file. The function’s coverage is now over 80%.

~ go tool cover -func=cover.out 
k8s.io/autoscaler/cluster-autoscaler/simulator/cluster.go:111:          NewRemovalSimulator                     100.0%
_**k8s.io/autoscaler/cluster-autoscaler/simulator/cluster.go:126:          SimulateNodeRemoval                     81.8%**_
k8s.io/autoscaler/cluster-autoscaler/simulator/cluster.go:168:          withForkedSnapshot                      66.7%
k8s.io/autoscaler/cluster-autoscaler/simulator/cluster.go:184:          findPlaceFor                            88.9%

**Comment by jackfrancis** [comment]
/lgtm
/approve
/hold cancel

thank you @kincoy!

**Comment by towca** [comment]
@kincoy @jackfrancis @x13n Do we know why the NodeInfo is missing in the first place? The fix itself looks good, but IIUC there was a time when this logic worked correctly even without the fix - do we know what changed? I'm worried the fix might be masking a wider regression in ClusterSnapshot.
